Source/bmalloc/libpas/src/test/EnumerationTests.cpp

```diff
@@ -0,0 +1,315 @@
+map<ReaderRange, void*> readerCache;
+
+void* enumeratorReader(pas_enumerator* enumerator,
+                       void* address,
+                       size_t size,
+                       void* arg)
+{
+    CHECK(!arg);
+    CHECK(size);
+
+    ReaderRange range = ReaderRange(address, size);
+
+    auto readerCacheIter = readerCache.find(range);
+    if (readerCacheIter != readerCache.end())
+        return readerCacheIter->second;
+
+    void* result = pas_enumerator_allocate(enumerator, size);
+
+    void* pageAddress = reinterpret_cast<void*>(
+        pas_round_down_to_power_of_2(
+            reinterpret_cast<uintptr_t>(address),
+            pas_page_malloc_alignment()));
+    size_t pagesSize =
+        pas_round_up_to_power_of_2(
+            reinterpret_cast<uintptr_t>(address) + size,
+            pas_page_malloc_alignment())
+        - reinterpret_cast<uintptr_t>(pageAddress);
+    void* pageEndAddress = reinterpret_cast<void*>(
+        reinterpret_cast<uintptr_t>(pageAddress) + pagesSize);
+
+    bool areProtectedPages = false;
+    if (!pageRanges.empty()) {
+        auto pageRangeIter = pageRanges.upper_bound(PageRange(pageAddress, pagesSize));
+        if (pageRangeIter != pageRanges.begin()) {
+            --pageRangeIter;
+            areProtectedPages =
+                pageRangeIter->base <= pageAddress
+                && pageRangeIter->end() > pageAddress;
+        }
+    }
+
+    if (verbose) {
+        cout << "address = " << address << "..."
+             << reinterpret_cast<void*>(reinterpret_cast<uintptr_t>(address) + size)
+             << ", pageAddress = " << pageAddress << "..." << pageEndAddress
+             << ", areProtectedPages = " << areProtectedPages << "\n";
+    }
+
+    if (areProtectedPages) {
+        int systemResult = mprotect(pageAddress, pagesSize, PROT_READ);
+        PAS_ASSERT(!systemResult);
+    }
+
+    memcpy(result, address, size);
+
+    if (areProtectedPages) {
+        int systemResult = mprotect(pageAddress, pagesSize, PROT_NONE);
+        PAS_ASSERT(!systemResult);
+    }
+
+    readerCache[range] = result;
+    return result;
+}
+
+void enumeratorRecorder(pas_enumerator* enumerator,
+                        void* address,
+                        size_t size,
+                        pas_enumerator_record_kind kind,
+                        void* arg)
+{
+    PAS_UNUSED_PARAM(enumerator);
+    CHECK(size);
+    CHECK(!arg);
+
+    if (verbose) {
+        cout << "Recording " << pas_enumerator_record_kind_get_string(kind) << ":" << address
+             << "..." << reinterpret_cast<void*>(reinterpret_cast<uintptr_t>(address) + size) << "\n";
+    }
+
+    RecordedRange range = RecordedRange(address, size);
+
+    CHECK(!recordedRanges[kind].count(range));
+    recordedRanges[kind].insert(range);
+}
+
+
+void testBasicEnumeration() {
+    if (verbose)
+        cout << "Testing basic enuemration" << endl;
+
+    /* create root */
```

Comment:
Nit: The "create root" comment doesn't really add much to the function call for `pas_root_create`

Reply:
Yeah, I removed these comments. They were more just some extra comments remaining around from development.